### ISSUE TITLE
Champion re-optimization under gap-aware fills (NQ + GC, all TFs)

### ISSUE THREAD

**Issue by mulhamfetna** [mulhamfetna/trading-strategy-finder#2]
Running now on the server (prefix `wshgap`), warm-started from each market's own champions so every result >= the prior champion re-scored honestly. Feeds the risk-budget re-cut.

Verify each new champion (dashboard-to-the-dollar, golden green) before it can be proposed as a champion via the champion-proposal template. See docs/superpowers/GAP-02.

**Comment by mulhamfetna** [comment]
⚠️ **Status correction — this run no longer exists.**

The issue says *"Running now on the server (prefix `wshgap`)"*. Verified today:

- **no `wshgap*` study, and no study matching `%gap%`, in the Optuna Postgres** (236 studies checked)
- **no optimizer process running**
- **no `wshgap` artifact** anywhere on the server or in the repo

So the re-optimization was lost — same as #14's `adoptgate_treat`, which also claimed to be running and had left nothing behind. Whatever the server was doing did not survive.

**What IS done and safe:** `docs/superpowers/GAP-02-champion-before-after.md` — the *measurement* of the gap-aware fill change across all 54 champion slots (P&L −0.2%, drawdown **+9.8%**, NG +148%). That finding stands.

**What is NOT done:** re-tuning the champions *for* the honest fill model. That is this issue, and it has not been started — or rather, it was started and is gone.

This blocks #3 (risk-budget re-cut), which is correctly labelled `status:blocked`.

**Before re-running it, note #14's finding:** if that re-optimization is launched with `--max-enabled`, it must use the fixed repair — otherwise it silently searches only 18 of the 165 indicators.

**Comment by mulhamfetna** [comment]
## Correcting my last comment: the run was **not** lost — and what it found does not hold up

My previous comment said this run had vanished and needed re-running from scratch. **That was wrong, and
I should not have concluded it from the absence of Optuna studies alone.** The re-optimization completed
on 2026-07-21/22 — 12/12 studies, no failures, 8h34m — and was written up in
`docs/superpowers/GAP-03-reoptimization-before-after.md`.

What was true is that **its outputs existed only as untracked files on the server**
(`~/Mulham/code/.worktrees/fundamental` and `~/Mulham/fa-m1/`). The Optuna studies really are gone —
0 of 238 in `wsh-pg` match `%gap%` — so those loose files were the only surviving record. They are now
committed at `optimize/reports/gap_fills/reopt_wshgap/` (8e4f2cf). This is exactly what
"LOCAL = source of truth" exists to prevent: for a week this issue read *"Running now on the server"*
while the only evidence sat on a box nobody was reading.

Having recovered it, I checked its conclusions rather than assuming them. **Four defects, in descending
order of how much they matter.**

### 1. It was measured against the wrong champion set — and the headline inverts

GAP-03 calls `wsh4_champions_full*.json` "deployed". It has not been deployed since **2026-07-14**, when
`DEFAULT_CHAMPION_SET` moved the deployed set to `best_*` (commit `4585648`) — eight days *before* this
run. `full_compare.py` loads `wsh4` as the baseline, so every "vs deployed" number in that report is
against a retired set.

Re-scored against the champions we actually run:

| | full-history | 2026 |
|---|---:|---:|
| GAP-03's claim (baseline = retired `wsh4`) | +$52,443 | **+$35,475** |
| **vs the truly deployed `best_*`** | +$94,522 | **−$12,832** |

Per adopted slot, against the real deployed champions:

| slot | deployed (`best_*`) | re-optimized | Δ |
|---|---:|---:|---:|
| NQ 1h | 27,203 | 38,008 | **+10,805** |
| GC 15m | 37,897 | 40,123 | **+2,226** |
| NQ 2h | 40,745 | 26,728 | **−14,017** |
| **net** | | | **−986** |

So the "three winners" are a **wash**, and **NQ 2h is $14,017 worse** than the champion it was meant to
replace. The gain was an artifact of the baseline, not an edge.

The evidence was already sitting on the server: `best_vs_wsh4.txt` was generated at **16:05 on
2026-07-22 — forty minutes *after* the adoption commit** — and nobody acted on it.

### 2. The search was floored against that same retired set (root cause)

`warm_start_seeds()` read `wsh4_champions_full*.json` **by name**. Optuna's warm start guarantees the
returned front scores ≥ its seed — and that guarantee never broke. It just held against the wrong
incumbent. That is why nothing looked wrong: every internal consistency check still passed.

This affected **every warm-started re-optimization run between 2026-07-14 and today**, not just this one.

Fixed in 8e4f2cf: the seed now resolves through `payload._instrument_champions_path()` — the same
resolver the dashboard uses — and each study **prints which file actually seeded it**, with a legacy
fallback announcing itself loudly instead of being taken silently.
`optimize/test_warm_start_seed_set.py` was verified to fail on the old code (2/2 assertions fire).

### 3. Its "2026 out-of-sample" column was not out-of-sample

`--train-window` did not exist in July — it was built last week for the adopt gate (#14). Without it the
walk-forward folds span the **whole** series, so 2026 was inside the optimizer's training data. GAP-03's
"2026 OOS" column is therefore contaminated: only the *seed* predated 2026, not the *search*.

Note the direction of that bias — it flatters 2026 results. The two configurations GAP-03 rejected as
over-fit (GC 4h, NQ 5m) failed **despite** the contamination, so those rejections stand.

### 4. Params persisted at 4 decimal places (provenance only, here)

The precision fix (`_sig`, 12 significant digits, `090d24b`) was not in the `fundamental-analysis`
worktree, so these were extracted by the old `round(x, 4)` path — the bug that flipped NG 5m's sign.
**For NQ and GC this is not materially distorting**: the smallest stop is GC 2m `sl_soft=2.2426`, still
5 significant figures (~0.002%). It would be fatal on NG or HG, so it matters for provenance and for any
future extension of this run.

### Nothing live was harmed

The adoption commit (`105a2da`) wrote into `wsh4_champions_full*.json` — the **retired** file — so the
deployed book was never actually changed. `wsh4` is not a registered champion set
(`test_champion_sets.py` forbids the prefix). One loose end: `presets.py` still reads `wsh4` directly for
the dashboard's preset dropdown, bypassing that registry check, and the golden baselines for 1h/2h were
re-captured to the adopted values (`96eb8de`) — so the gate is self-consistent with `wsh4` rather than
with what is deployed.

GAP-03 has been marked RETRACTED with all of the above recorded inline.

### Next: re-running it properly

Seeded from `best_*` (fix in place), on the current engine, **`--train-window 2025` so 2026 is a genuine
holdout**, new prefix `wshgap2`. Note the engine changed materially after the July run — #62 (indicator
acceleration), #74, and #75 (which fixed cross-series wiring outright) — so those numbers would not
reproduce today even with correct seeding.

I will run the **gate question first**: train on 2025, judge on the untouched 2026 year, and ask whether
re-tuning under honest fills beats the deployed champion *at all*. Only if it does is a full-history run
for deployment worth another 8 hours.

### ⚠️ #3 is probably not blocked on this

#3 says "re-derive on the new champions' drawdowns", blocked on this issue. Two reasons that dependency
looks weaker than written:

1. **Scope mismatch.** #3 needs a portfolio risk budget across the whole book — **54 slots**. This issue
   covers **12** (NQ + GC). It could never have fed #3 on its own.
2. **The input #3 actually needs already exists.** GAP-02 measured honest-fill drawdowns across all 54
   deployed champions (P&L −0.2%, drawdown **+9.8%**, NG **+148%**). That measurement stands and is
   untouched by any of the above.

If the re-run changes no champions — which the numbers above make likely — #3 can proceed immediately on
GAP-02's drawdowns. I'd suggest unblocking it on that basis rather than waiting.

**Comment by mulhamfetna** [comment]
## Correction: the July studies are **not** gone either — I searched the wrong database

In my last comment I said *"The Optuna studies really are gone — 0 of 238 in `wsh-pg` match `%gap%`."*
**That is wrong.**

The optimizer writes to Postgres **only when `WSH_STORAGE_URL` is set**. Neither the July campaign nor
the current one sets it, so studies go to **per-timeframe SQLite files** under
`optimize/studies/wsh_<tf>[_<INST>].db`. I queried the Postgres instance and found nothing, which was
true and irrelevant.

All 12 July studies are intact in the `fundamental` worktree, with full trial histories:

```
wsh_4h.db      wshgap_4h       trials=6142      wsh_4h_GC.db   wshgap_4h_GC    trials=5900
wsh_2h.db      wshgap_2h       trials=5900      wsh_2h_GC.db   wshgap_2h_GC    trials=5900
wsh_1h.db      wshgap_1h       trials=5900      wsh_1h_GC.db   wshgap_1h_GC    trials=5900
wsh_15m.db     wshgap_15m      trials=5900      wsh_15m_GC.db  wshgap_15m_GC   trials=5900
wsh_5m.db      wshgap_5m       trials=5900      wsh_5m_GC.db   wshgap_5m_GC    trials=5900
wsh_2m.db      wshgap_2m       trials=5900      wsh_2m_GC.db   wshgap_2m_GC    trials=5900
```

**What this changes:** July's run can be **re-extracted at full precision** — now that `_exact` has
replaced the 4-dp persistence path — without re-running anything. That removes defect 4 from my earlier
list at zero cost, if we ever want those champions re-derived cleanly.

**What this does not change:** the run still should not be adopted. It was warm-started from the retired
`wsh4_*` set, compared against that same wrong baseline (**−$12,832 out-of-sample vs the champions we
actually run**), and its "2026 out-of-sample" year was inside its own training data. The corrected
re-run (`wshgap4`) remains the right call and is in progress — NQ 4h finished in 35 min, seeded from
`best_champions_full.json` as intended.

**The lesson, since this is now twice on the same issue:** I twice concluded "gone" from a search that
came back empty. An empty result means *"not here"* — it never means *"nowhere"*. Both times the correct
next step was to check where the data is actually written, not to trust the absence.

Docs corrected: `optimize/reports/gap_fills/reopt_wshgap/README.md` and
`docs/reports-2026-07-29/ISSUE-02-champion-reoptimization.md`.

**Comment by mulhamfetna** [comment]
# GAP-04 — Honest-fill re-optimization, judged on a real holdout: **ADOPT NOTHING**

**Date:** 2026-07-30 · **Issue:** #2 · **Study prefix:** `wshgap4`
**Supersedes:** [GAP-03](./GAP-03-reoptimization-before-after.md) (retracted — wrong baseline, contaminated holdout)

---

## Verdict

**No champion is adopted. The deployed set wins on every slot that moved.**

| | train 2025 (what the optimizer saw) | **holdout 2026 (never seen)** |
|---|---:|---:|
| deployed (`best_*`) | $501,126 | **$309,072** |
| re-optimized (`wshgap4`) | **$761,472** | **$67,419** |
| delta | **+$260,346** | **−$241,653** |

| slots better on the holdout | **0** |
|---|---|
| slots worse | **9** |
| slots unchanged (warm start could not be beaten) | 3 |
| re-optimized slots that turn **negative** in 2026 | **6** |

The re-optimization gained **+52% on the year it was allowed to look at** and lost **−78% on the year it
was not**. That is textbook over-fitting, and it is visible only because this run finally had a genuine
holdout.

## The full table

| slot | train 2025 | holdout 2026 | holdout drawdown |
|---|---|---|---|
| NQ 4h | 90,054 → 90,054 | 61,601 → 61,601 | 15,560 → 15,560 |
| NQ 2h | 41,657 → 97,898 | 40,745 → **−21,830** | 5,645 → 24,266 |
| NQ 1h | 56,620 → 75,245 | 27,203 → 5,040 | 7,825 → 10,784 |
| NQ 15m | 22,515 → 29,835 | 21,306 → **−298** | 4,939 → 2,853 |
| NQ 5m | 13,091 → 21,109 | 2,600 → **−343** | 1,521 → 3,295 |
| NQ 2m | 26,010 → 25,608 | 4,545 → **−5,003** | 3,833 → 9,993 |
| GC 4h | 54,782 → 121,292 | 34,290 → **−16,090** | 7,969 → 16,178 |
| GC 2h | 58,828 → 134,191 | 26,101 → **−20,524** | 15,649 → 24,863 |
| GC 1h | 58,228 → 88,632 | 28,740 → 12,301 | 21,007 → 21,091 |
| GC 15m | 44,713 → 44,713 | 37,897 → 37,897 | 7,583 → 7,583 |
| GC 5m | 9,619 → 7,884 | 10,346 → 971 | 3,905 → 1,116 |
| GC 2m | 25,008 → 25,008 | 13,699 → 13,699 | 3,825 → 3,825 |

Look at **GC 2h**: training profit more than doubled ($58,828 → $134,191) while the holdout went from
+$26,101 to **−$20,524** and drawdown grew 59%. Every ingredient of a strategy that has memorised its
training year.

## ⚠️ What this does and does not prove

**It proves:** none of these twelve re-optimized champions should be adopted. That decision is solid —
zero of them improve the untouched year, and half of them lose money in it.

**It does NOT prove that re-optimizing under honest fills is worthless**, and the reason is a limitation
of my own design, not of the result:

> The deployed champions were tuned on the **full history**. `wshgap4` was restricted to **2025 only**
> (`--train-window 2025`), because that is what creates a genuine 2026 holdout. So the comparison
> confounds two changes: *honest fills* and *a training window less than half as long*.


---

Full report: `docs/superpowers/GAP-04-reopt-holdout-verdict.md` (17a308e).
Artifacts: `optimize/reports/gap_fills/reopt_wshgap4/` — champions, comparison JSON, run logs, and the
three scripts, all committed locally so this run cannot go missing the way July's did.

**Closing.** The question this issue asked — *"if we re-tune the champions honestly, do we get better
strategies?"* — is answered for the configuration tested: **no, and not marginally.** The deployed set
stands unchanged.

**Comment by mulhamfetna** [comment]
## ⛔ Correction to the verdict I posted hours ago — the comparison was not symmetric

Found while checking the data range before starting the follow-up work. **The decision (adopt nothing)
survives. The headline number does not.**

### The error

`optimize/folds.py: score_walkforward` calls `split_folds(df_dec, k)` over the **whole** series. So the
deployed `best_*` champions were *selected* using folds that **include 2026**, while `wshgap4` was
truncated to 2025 by `--train-window 2025`.

> The "holdout 2026" column is therefore **in-sample for the deployed set and out-of-sample for the
> challenger.** I built a genuine holdout for the candidate and never checked whether the incumbent had
> one — which is precisely the trap #14 exists to document, reproduced inside the experiment designed to
> avoid it.

**So −$241,653 / −78% is not a fair measure of what re-optimization is worth.** It scores an
out-of-sample result against an in-sample one.

### A second, smaller error

I wrote that `wshgap4` had "less than half the training data". It had **72%**:

| | decision bars | share |
|---|---:|---:|
| total (2025-01-01 → 2026-05-19) | 2,119 | 100% |
| `wshgap4` training (2025) | 1,534 | **72.4%** |
| "holdout" 2026 | 585 | 27.6% — **138 days** |

### What stands, and why

**Adopt nothing is unchanged**, because it rests on an absolute criterion that does not involve the
baseline:

> **6 of 12 re-optimized champions lose money on data they never saw** — NQ 2h −$21,830, GC 2h −$20,524,
> GC 4h −$16,090, NQ 2m −$5,003, NQ 5m −$343, NQ 15m −$298. And **0 of 12 improved** the unseen year.

A strategy that is negative out-of-sample is not adoptable whatever it is compared against. Issue stays
closed.

### The bigger thing this exposes

We have **1.38 years of history**, and the deployed champions have effectively seen **all of it**. So
there is **no window on which a fair head-to-head can be run** with this dataset. That is not a flaw in
the experimental design — it is a limit of the data.

**#87 is re-scoped accordingly:** the blocker is *history length*, not the choice of training window. A
"longer training window with a later holdout" is not available to us; either we extend the price history
before 2025, or the question stays unanswerable.

GAP-04 has been updated with all of the above, including what should have prompted the check earlier: a
"holdout year" only 138 days long.
